editor/plugins/texture_editor_plugin.cpp

```diff
@@ -124,19 +131,48 @@ void TexturePreview::_update_metadata_label_text() {
 }
 
 TexturePreview::TexturePreview(Ref<Texture2D> p_texture, bool p_show_metadata) {
-	checkerboard = memnew(TextureRect);
-	checkerboard->set_stretch_mode(TextureRect::STRETCH_TILE);
-	checkerboard->set_texture_repeat(CanvasItem::TEXTURE_REPEAT_ENABLED);
-	checkerboard->set_custom_minimum_size(Size2(0.0, 256.0) * EDSCALE);
-	add_child(checkerboard);
+	set_custom_minimum_size(Size2(0.0, 256.0) * EDSCALE);
+
+	bg_rect = memnew(ColorRect);
+	bg_rect->set_anchors_preset(PRESET_FULL_RECT);
+	bg_rect->set_clip_contents(true);
+	add_child(bg_rect);
+
+	MarginContainer *margin_container = memnew(MarginContainer);
+	margin_container->set_anchors_preset(PRESET_FULL_RECT);
+	margin_container->add_theme_constant_override("margin_right", 1);
+	margin_container->add_theme_constant_override("margin_top", 1);
+	margin_container->add_theme_constant_override("margin_left", 1);
+	margin_container->add_theme_constant_override("margin_bottom", 1);
+	add_child(margin_container);
+
+	centering_container = memnew(AspectRatioContainer);
+	centering_container->set_anchors_preset(PRESET_FULL_RECT);
+	margin_container->add_child(centering_container);
 
 	texture_display = memnew(TextureRect);
 	texture_display->set_texture_filter(TEXTURE_FILTER_NEAREST_WITH_MIPMAPS);
 	texture_display->set_texture(p_texture);
-	texture_display->set_anchors_preset(TextureRect::PRESET_FULL_RECT);
-	texture_display->set_stretch_mode(TextureRect::STRETCH_KEEP_ASPECT_CENTERED);
 	texture_display->set_expand_mode(TextureRect::EXPAND_IGNORE_SIZE);
-	add_child(texture_display);
+	texture_display->set_anchors_preset(PRESET_FULL_RECT);
+	centering_container->add_child(texture_display);
+
+	checkerboard = memnew(TextureRect);
+	checkerboard->set_anchors_preset(PRESET_FULL_RECT);
+	checkerboard->set_expand_mode(TextureRect::EXPAND_IGNORE_SIZE);
+	checkerboard->set_stretch_mode(TextureRect::STRETCH_TILE);
+	checkerboard->set_texture_repeat(CanvasItem::TEXTURE_REPEAT_ENABLED);
+	checkerboard->set_draw_behind_parent(true);
+	texture_display->add_child(checkerboard);
+
+	borders_rect = memnew(ReferenceRect);
+	borders_rect->set_border_width(2);
+	borders_rect->set_anchors_preset(PRESET_FULL_RECT);
+	borders_rect->set_draw_behind_parent(true);
+	checkerboard->add_child(borders_rect);
```

Comment:
Hmm, actually half of this 2-width border will be behind the texture display area (ReferenceRect draws the border with its center being on the rect edge), meaning it could add some visual noise in case the overlapping checkerboard's / the displayed texture's border pixel colors are semi-transparent.

So correct would be to draw only the 1-width border part outside the texture display area, so no overlapping could happen. The question is how exactly to achieve this. AFAICT can't use MarginContainer(s) for 0.5 offsets as they support only integer margins. One way would be to replace this ReferenceRect with a plain Control and manually draw needed rect on its `draw` signal, like `border_drawer.draw_rect(Vector2(-0.5, -0.5), border_drawer.get_size() + Vector2(1, 1), border_color, false, 1.0);`

Alternative could be e.g. adding different border placement mode(s) to ReferenceRect so you could pick to draw the border inside the rect instead of on its edge (then you could just change such mode and set the width to `1`). But this might be overdoing/overcomplicating, not sure.

---

BTW note that the hierarchy could be flatten a little (so all fitting-in would be done directly by the containers, and there would be no need for `draw_behind_parent`), like:

| Current | More flat |
|--------|--------|
|![8c8pHDaBji](https://github.com/user-attachments/assets/6b2c28ff-71f9-4ae2-af18-e612fdf3c0f2)|![FHvGNszGyu](https://github.com/user-attachments/assets/99480c19-a2d7-41da-8f92-ff7d4f46d731)|

Seems simpler to me, but probably fine as is / not a big deal. :upside_down_face:

Reply:
Yes, this may happen because checkerboard texture is semitransparent. But from what I saw this overlapping doesn't bring much noise.
I'll try to replace `ReferenceRect` with manual drawing, maybe it will look better (or at least comparable).

About tree flattering - goo point, will change it.

Reply:
Also I found out that `TexturePreview` is used for `Camera3DPreview` which doesn't set texture on creation and current implementation will crash editor on `Camera3D` selection 💀

Reply:
I have removed `ReferenceRect` and connected outline drawing to `TextureRect`.
Also outline width now could be configured.